instrumentation/netty/netty-4.0/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_0/ChannelPipelineTest.java

```diff
@@ -0,0 +1,219 @@
+    assertNull(channelPipeline.last());
+    assertEquals(0, channelPipeline.toMap().size());
+
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addFirst("test", noopHandler);
+
+    // only the noop handler
+    assertEquals(noopHandler, channelPipeline.first());
+    assertEquals(0, channelPipeline.toMap().size());
+
+    if ("by instance".equals(desc)) {
+      channelPipeline.replace(noopHandler, "http", httpHandler);
+    } else if ("by class".equals(desc)) {
+      channelPipeline.replace(noopHandler.getClass(), "http", httpHandler);
+    } else if ("by name".equals(desc)) {
+      channelPipeline.replace("test", "http", httpHandler);
+    }
+
+    // noop handler was removed; http and instrumentation handlers were added
+    assertEquals(httpHandler, channelPipeline.first());
+    assertEquals("HttpClientTracingHandler", channelPipeline.last().getClass().getSimpleName());
+    assertEquals(1, channelPipeline.toMap().size());
+
+    NoopChannelHandler anotherNoopHandler = new NoopChannelHandler();
+    channelPipeline.replace("http", "test", anotherNoopHandler);
+
+    // http and instrumentation handlers were removed; noop handler was added
+    assertEquals(anotherNoopHandler, channelPipeline.first());
+  }
+
+  // regression test for
+  // https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/4056
+  @Test
+  void shouldAddAfterAndRemoveLastHandler() throws Exception {
+    EmbeddedChannel channel = new EmbeddedChannel(new NoopChannelHandler());
+    ChannelPipeline channelPipeline = (ChannelPipeline) getConstructor().newInstance(channel);
+    HttpClientCodec httpHandler = new HttpClientCodec();
+
+    // no handlers initially
+    assertEquals(
+        "io.netty.channel.DefaultChannelPipeline$TailHandler",
+        channelPipeline.first().getClass().getName());
+    assertNull(channelPipeline.last());
+    assertEquals(0, channelPipeline.toMap().size());
+
+    // Add http and instrumentation handlers
+    channelPipeline.addLast("http", httpHandler);
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals("HttpClientTracingHandler", channelPipeline.last().getClass().getSimpleName());
+    assertEquals(1, channelPipeline.toMap().size());
+
+    NoopChannelHandler noopHandler = new NoopChannelHandler();
+    channelPipeline.addAfter("http", "noop", noopHandler);
+
+    // instrumentation handler is between http and noop handlers
+    assertEquals(channelPipeline.first(), httpHandler);
+    assertEquals(channelPipeline.last(), noopHandler);
+    assertEquals(1, channelPipeline.toMap().size());
+
+    ChannelHandler removed = channelPipeline.removeLast();
+    assertEquals(httpHandler, removed);
+    assertEquals(0, channelPipeline.toMap().size());
+
+    // http and instrumentation handlers will be remained; noop handler will be removed
+    // but in reality, line 162 will remove http, instrumentation, and noop handlers
```

Comment:
please read my comments here.

Reply:
The issue is in something else. Also note that using line numbers in the comment that way isn't the best idea as the line numbers can easily change.